src/attribute_collector.hpp

```diff
@@ -0,0 +1,100 @@
+
+#pragma once
+
+#include <span>
+#include <string_view>
+#include <unordered_map>
+
+#include "object_store.hpp"
+#include "target_address.hpp"
+#include "utils.hpp"
+
+namespace ddwaf {
+
+class attribute_collector {
+public:
+    attribute_collector() { ddwaf_object_map(&attributes_); }
+    attribute_collector(const attribute_collector &) = delete;
+    attribute_collector &operator=(const attribute_collector &) = delete;
+    attribute_collector(attribute_collector &&other) noexcept = delete;
+    attribute_collector &operator=(attribute_collector &&other) noexcept = delete;
+    ~attribute_collector() { ddwaf_object_free(&attributes_); }
+
+    // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    bool emplace(std::string_view key, std::string_view value)
+    {
+        ddwaf_object object;
+        ddwaf_object_stringl(&object, value.data(), value.size());
+        return emplace(key, object, false);
+    }
+
+    bool emplace(std::string_view key, uint64_t value)
+    {
+        ddwaf_object object;
+        ddwaf_object_unsigned(&object, value);
+        return emplace(key, object, false);
+    }
+
+    bool emplace(std::string_view key, int64_t value)
+    {
+        ddwaf_object object;
+        ddwaf_object_signed(&object, value);
+        return emplace(key, object, false);
+    }
+
+    bool emplace(std::string_view key, double value)
+    {
+        ddwaf_object object;
+        ddwaf_object_float(&object, value);
+        return emplace(key, object, false);
+    }
+
+    bool emplace(std::string_view key, bool value)
+    {
+        ddwaf_object object;
+        ddwaf_object_bool(&object, value);
+        return emplace(key, object, false);
+    }
+
+    bool emplace(std::string_view key, const ddwaf_object &value, bool copy);
+
+    bool collect(const object_store &store, target_index input_target,
+        std::span<const std::string> input_key_path, std::string_view output);
+
+    void collect_pending(const object_store &store);
+
+    [[nodiscard]] bool has_pending_attributes() const { return !pending_.empty(); }
+
+    ddwaf_object get_available_attributes()
+    {
+        auto output_object = attributes_;
+        // Reset attributes
+        ddwaf_object_map(&attributes_);
+        emplaced_or_pending_attributes_.clear();
+
+        return output_object;
+    }
+
+protected:
+    enum class collection_state : uint8_t { success, unavailable, failed };
+
+    collection_state collect_helper(const object_store &store, target_index input_target,
+        std::span<const std::string> input_key_path, std::string_view output);
+    bool emplace_helper(std::string_view key, const ddwaf_object &value, bool copy);
+
+    // The views and spans used here are owned by rules and processors, these
+    // are part of their definition and are unchanging during the lifetime of
+    // the context, therefore it's safe to use them.
+    using target_type = std::pair<target_index, std::span<const std::string>>;
```

Comment:
wouldn't hurt to use a struct or at least mention what the span is about

Reply:
Added a comment explaining the purpose of both elements.